kernels/EvalKernel.cpp

```diff
@@ -0,0 +1,169 @@
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+ * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+ * OF SUCH DAMAGE.
+ ****************************************************************************/
+
+#include "EvalKernel.hpp"
+
+#include <pdal/KDIndex.hpp>
+#include <pdal/PDALUtils.hpp>
+#include <pdal/pdal_config.hpp>
+#include <pdal/util/ProgramArgs.hpp>
+
+namespace pdal
+{
+
+using namespace Dimension;
+
+static StaticPluginInfo const s_info{
+    "kernels.eval", "Eval Kernel", "http://pdal.io/kernels/kernels.eval.html"};
+
+CREATE_STATIC_KERNEL(EvalKernel, s_info)
+
+std::string EvalKernel::getName() const
+{
+    return s_info.name;
+}
+
+void EvalKernel::addSwitches(ProgramArgs& args)
+{
+    args.add("predicted", "Point cloud filename containing predicted labels",
+             m_predictedFile)
+        .setPositional();
+    args.add("truth", "Point cloud filename containing truth labels",
+             m_truthFile)
+        .setPositional();
+    args.add("labels",
+             "Comma-separated list of classification labels to evaluate",
+             m_labelStrList);
+}
+
+void EvalKernel::validateSwitches(ProgramArgs& args)
+{
+    if (m_labelStrList.empty())
+        throw pdal_error(
+            "Must specify comma-separated list of labels to evaluate.");
+}
+
+PointViewPtr EvalKernel::loadSet(const std::string& filename,
+                                 PointTableRef table)
+{
+    Stage& reader = makeReader(filename, "");
+    reader.prepare(table);
+    PointViewSet viewSet = reader.execute(table);
+    assert(viewSet.size() == 1);
+    return *viewSet.begin();
+}
+
+int EvalKernel::execute()
+{
+    ColumnPointTable predictedTable;
+    PointViewPtr predictedView = loadSet(m_predictedFile, predictedTable);
+
+    ColumnPointTable truthTable;
+    PointViewPtr truthView = loadSet(m_truthFile, truthTable);
+
+    assert(predictedView->size() == truthView->size());
+
+    KD3Index& kdi = truthView->build3dIndex();
+
+    int dim = m_labelStrList.size();
+
+    std::vector<int> labelList;
+    for (auto const& label : m_labelStrList)
+        labelList.push_back(std::stoi(label));
+    std::sort(labelList.begin(), labelList.end());
+
+    LabelStats ls(dim);
+
+    for (PointRef p : *predictedView)
+    {
+        // It would be nice if we could expect that the points are aligned in
+        // both the predicted and truth views, but this often cannot be
+        // guaranteed, so rather than using the same PointId, we search for the
+        // nearest neighbor.
+        PointId qid = kdi.neighbor(p);
+        PointRef q = truthView->point(qid);
+
+        // TODO (chambbj): We should perhaps look at the distance to the
+        // nearest point and reject or otherwise report distances greater than
+        // 0.0, indicating some sort of mismatch between files.
+
+        int pc = p.getFieldAs<int>(Id::Classification);
```

Comment:
Maybe add an option to default to `Classification` so that a user could compare to their own dimensions other than `Classification` for whatever reason.

Reply:
In my mind, I didn't want it to expand too far beyond comparing classifications, but I hadn't considered it could be useful to compare classifications in a custom dimension, which seems reasonable. I can make this change. (I also realize that it _could_ be handy to compare two separate dimensions in a single file, which even makes this streamable and cuts down on IO. I'll have to think about whether that is a tweak to this kernel or something different altogether. Also need to think about how useful that is as a general purpose tool, or if it's just too specialized.)